spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt

```diff
@@ -6,182 +6,174 @@ import org.jetbrains.exposed.sql.StdOutSqlLogger
 import org.jetbrains.exposed.sql.Transaction
 import org.jetbrains.exposed.sql.addLogger
 import org.jetbrains.exposed.sql.exposedLogger
-import org.jetbrains.exposed.sql.statements.api.ExposedConnection
-import org.jetbrains.exposed.sql.statements.jdbc.JdbcConnectionImpl
-import org.jetbrains.exposed.sql.transactions.TransactionInterface
 import org.jetbrains.exposed.sql.transactions.TransactionManager
-import org.springframework.jdbc.datasource.ConnectionHolder
-import org.springframework.jdbc.datasource.DataSourceTransactionManager
+import org.jetbrains.exposed.sql.transactions.transactionManager
 import org.springframework.transaction.TransactionDefinition
 import org.springframework.transaction.TransactionSystemException
-import org.springframework.transaction.support.DefaultTransactionDefinition
+import org.springframework.transaction.support.AbstractPlatformTransactionManager
 import org.springframework.transaction.support.DefaultTransactionStatus
-import org.springframework.transaction.support.TransactionSynchronizationManager
+import org.springframework.transaction.support.SmartTransactionObject
 import javax.sql.DataSource
 
 class SpringTransactionManager(
     dataSource: DataSource,
-    databaseConfig: DatabaseConfig = DatabaseConfig { },
+    databaseConfig: DatabaseConfig = DatabaseConfig {},
     private val showSql: Boolean = false,
-    @Volatile override var defaultReadOnly: Boolean = databaseConfig.defaultReadOnly,
-    @Volatile override var defaultRepetitionAttempts: Int = databaseConfig.defaultRepetitionAttempts,
-    @Volatile override var defaultMinRepetitionDelay: Long = databaseConfig.defaultMinRepetitionDelay,
-    @Volatile override var defaultMaxRepetitionDelay: Long = databaseConfig.defaultMaxRepetitionDelay
-) : DataSourceTransactionManager(dataSource), TransactionManager {
+) : AbstractPlatformTransactionManager() {
```

Comment:
A question about breaking changes and migration for users since `TransactionManager` interface will no longer be implemented:

- All interface functions will no longer be callable on a `SpringTransactionManager` instance, namely `currentOrNull()` and `newTransaction()`:
    - For example, the user in [this issue](https://github.com/JetBrains/Exposed/issues/1673) uses `newTransaction()` to set a non-default isolation level. Is the plan that this use would be replaced with `@Transactional(isolation = ?)`?
    - And would it be possible to return the current transaction, by the user going through the database, like `tm.database.transactionManager.currentOrNull()`? Or is the intention that this implementation delegates all transaction operations to Spring, so the user should not need to call these functions anymore?

Reply:
I didn't realize there was a case where a user could create and use a new instance object that wasn't a bean.

If used in this way, I think it's possible that Spring's transaction management and Exposed's transaction management could get mixed up, making it difficult to manage connections.

1. I would prefer not to use newTransaction and currentOrNull in a SpringTransactionManager instance in this situation, and if you agree, it would be a good idea to implement a TransactionManager that throws an error if newTransaction and currentOrNull are used.

2. the code provides an isolation level by using @Transactional(isolation = ?), but I haven't verified that this properly implements Spring's policy. I'll try to implement Isolation and Propagation as per Spring's policy in the following PR.
```kotlin
    override fun doBegin(transaction: Any, definition: TransactionDefinition) {
        val trxObject = transaction as ExposedTransactionObject

        val currentTransactionManager = trxObject.manager
        TransactionManager.resetCurrent(currentTransactionManager)

        currentTransactionManager.currentOrNull()
            ?: currentTransactionManager.newTransaction(
                isolation = definition.isolationLevel,
                readOnly = definition.isReadOnly,
            ).apply {
                if (showSql) {
                    addLogger(StdOutSqlLogger)
                }
            }
    }
```

3. I exposed the database to check if the TransactionManager is set up properly when testing, but if there is another way to check the database at test time, I will prevent the database instance from being accessed with tm.database. Can you give me any advice?

Reply:
1. For the next release, the loss of `TransactionManager` functions would be mentioned as a breaking change and I'm drafting a migration section to make clear any changes. For now we'll see what issues pop up for example if users choose to call a new transaction within a `SpringTransactionManager` instance in spite of our suggestions.
2. Sounds good to me.
3. Our core unit tests usually involve passing in the database directly so it is available as a test property. But in special cases, it can be checked via `TransactionManager.currentOrNull()?.db` (see [MultidatabaseTest.kt](https://github.com/JetBrains/Exposed/blob/main/exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/h2/MultiDatabaseTest.kt#L38) for an example). I ran your tests using the following assertion and they all passed so it could be a valid replacement for `tm.database`:
```kt
assertEquals(
    TransactionManager.managerFor(this.database),
    TransactionManager.managerFor(TransactionManager.currentOrNull()?.db)
)
assertEquals(
    TransactionManager.managerFor(TransactionManager.currentOrNull()?.db),
    TransactionManager.manager
)
```
Note that the tests that pre-assign a variable would need to call this database check instead only when needed:
```kt
val tm = SpringTransactionManager(ds1)
val tm2 = SpringTransactionManager(ds2)
// val transactionManager2 = TransactionManager.managerFor(tm2.database)

tm2.executeAssert(false) {
    tm.executeAssert(false)
    assertEquals(TransactionManager.managerFor(TransactionManager.currentOrNull()?.db), TransactionManager.manager)
}
```

Reply:
> And would it be possible to return the current transaction, by the user going through the database, like tm.database.transactionManager.currentOrNull()? Or is the intention that this implementation delegates all transaction operations to Spring, so the user should not need to call these functions anymore?

I changed the database instance to be inaccessible directly, and the transaction is now only accessible via TransactionManager.currentOrNull(). And my intention is, as you say, that the user should not be able to use the Exposed Transaction directly when using the PlatformTransactionManager Bean. 

However, even if the user accesses the transaction with TransactionManager.currentOrNull(), it will behave almost identically to the transaction { } in Exposed, so as long as you don't call transaction.close(), you shouldn't have any problems.

The recommendation is to avoid calling the TransactionManager directly.